docs/design.md

```diff
@@ -0,0 +1,785 @@
+# Deckhand Design
+
+## Purpose
+
+Deckhand is a document-based configuration storage service built with
+auditability and validation in mind.
+
+## Essential Functionality
+
+* layering - helps reduce duplication in configuration while maintaining
+  auditability across many sites
+* substitution - provides separation between secret data and other
+  configuration data, while allowing a simple interface for clients
+* revision history - improves auditability and enables services to provide
+  functional validation of a well-defined collection of documents that are
+  meant to operate together
+* validation - allows services to implement and register different kinds of
+  validations and report errors
+
+## Documents
+
+All configuration data is stored entirely as structured documents, for which
+schemas must be registered.
+
+### Document Format
+
+The document format is modeled loosely after Kubernetes practices. The top
+level of each document is a dictionary with 3 keys: `schema`, `metadata`, and
+`data`.
+
+* `schema` - Defines the name of the JSON schema to be used for validation.
+  Must have the form: `<namespace>/<kind>/<version>`, where the meaning of
+  each component is:
+  * `namespace` - Identifies the owner of this type of document. The
+    values `deckhand` and `metadata` are reserved for internal use.
+  * `kind` - Identifies a type of configuration resource in the namespace.
+  * `version` - Describe the version of this resource, e.g. "v1".
+* `metadata` - Defines details that Deckhand will inspect and understand. There
+  are multiple schemas for this section as discussed below. All the various
+  types of metadata include a `name` field which must be unique for each
+  document `schema`.
+* `data` - Data to be validated by the schema described by the `schema`
+  field. Deckhand only interacts with content here as instructed to do so by
+  the `metadata` section. The form of this section is considered to be
+  completely owned by the `namespace` in the `schema`.
+
+#### Document Metadata
+
+There are 3 supported kinds of document metadata. Documents with `Document`
+metadata are the most common, and are used for normal configuration data.
+Documents with `Control` metadata are used to customize the behavior of
+Deckhand. Documents with `Tombstone` metadata are used to delete pre-existing
+documents with either `Document` or `Control` metadata.
+
+##### schema: metadata/Document/v1
+
+This type of metadata allows the following metadata hierarchy:
+
+* `name` - string, required - Unique within a revision for a given `schema`.
+* `storagePolicy` - string, required - Either `cleartext` or `encrypted`. If
+  `encyrpted` is specified, then the `data` section of the document will be
```

Comment:
What I'm referring to is that deckhand authorization policy. A per-document may be overly restrictive - but as it stands it would appear that a service would be able to request credentials and keys that have nothing to do with it - as deckhand - though storing secrets securely, would appear to hand them out to anyone who requested them.